### PR TITLE
[upstreamca/disk] support joining existing PKI

### PR DESCRIPTION
**Pull request by ZymoticB**
Signed-off-by: Tyler Dixon <tylerd@uber.com>

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
This PR updates the server "disk" upstreamca plugin to support joning existing
PKI by loading a cert chain and trust bundle.

I'm not a big fan of the mode where the trust bundle is implicitly the last cert
in the chain but it gives a sane default (rather than failing to load the CA if
bundle_file_path is unset and the cert_file_path contains more than one cert) and
it is potentially convenient.

### CODE REVIEW

File: pkg/server/plugin/upstreamca/disk/disk.go [hunk 1]
```diff
@@ -108,57 +113,84 @@ func (p *DiskPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCS
...
+
+	var trustBundle []byte
+	if config.BundleFilePath == "" {
+		// If there is no bundle path configured then we assume the chain
+		// including the root comes from cert_file_path. The final cert
+		// is the root that should be trusted. Additionally, it can be dropped
+		// from the cert chain to avoid wastefully distributing it in the chain
+		// of an SVID as well as the trust bundle.
+		trustBundle = certs[len(certs)-1].Raw
+		certs = certs[:len(certs)-1]
```
Comment (APTy) [inline pkg/server/plugin/upstreamca/disk/disk.go]: i think we want to be careful about using a cert chain here to represent the upstream ca, and may want to put as much as possible into the trust bundle

from my reading, if you're joining an upstream ca pki (which i believe we are), the rest of the cert chain is tossed out

https://github.com/spiffe/spire/blob/5ef1c258bb27e0abf1ef0235e60e610ac07506e5/pkg/server/ca/manager.go#L967-L971

Comment (azdagron) [reply pkg/server/plugin/upstreamca/disk/disk.go]: That isn't the case. `caChain` contains the `CertChain` returned by the upstream CA. When you join upstream PKI, `upstreamChain` is set to `caChain` (preserving the whole chain returned by the upstream CA). `caChain[0]` is the server CA certificate signed by upstream.

Comment (APTy) [reply pkg/server/plugin/upstreamca/disk/disk.go]: Ah I see now. So is the full upstreamChain appended to the leaf SVID then?

Comment (ZymoticB) [reply pkg/server/plugin/upstreamca/disk/disk.go]: Yes https://github.com/spiffe/spire/blob/master/pkg/server/ca/ca.go#L291

**Comment by azdagron** [comment]
Grabbing the last cert in the cert file as the "trust bundle" seems just a little dangerous, in that it could trivially mask a misconfiguration (where `bundle_file_path` was accidentally omitted). I'm not entirely convinced it is worth the convenience of being able to distribute a single file.

If we do keep that feature, I think we should add a check that the last certificate is self-signed.
Alternatively we should fail Configure() if the cert file contains more than one certificate and the `bundle_file_path` is unset.

It might also be worth adding a check that the chain+trust bundle forms a valid chain of trust (via a call to (*x509.Certificate).Verify()).

**Comment by ZymoticB** [comment]
> Grabbing the last cert in the cert file as the "trust bundle" seems just a little dangerous, in that it could trivially mask a misconfiguration (where `bundle_file_path` was accidentally omitted). I'm not entirely convinced it is worth the convenience of being able to distribute a single file.
> 
> If we do keep that feature, I think we should add a check that the last certificate is self-signed.
> Alternatively we should fail Configure() if the cert file contains more than one certificate and the `bundle_file_path` is unset.
> 
> It might also be worth adding a check that the chain+trust bundle forms a valid chain of trust (via a call to (*x509.Certificate).Verify()).

Yes, self-validation is missing. I will add that.

I will fail Configure if there is more than one certificate and no bundle. Thanks for the second opinion.

File: pkg/server/plugin/upstreamca/disk/disk_test.go [hunk 2]
```diff
@@ -43,6 +44,11 @@ type DiskSuite struct {
 func (s *DiskSuite) SetupTest() {
 	p := New()
 
+	// This ensures that there are only specific tests that do the verify
+	// flow lowering the cost of "refreshing" all of the cert material
+	// associated with the tests in this package. TODO before 2029 generate
```
Comment (azdagron) [inline pkg/server/plugin/upstreamca/disk/disk_test.go]: I'm hoping we get to this sooner than later, using certs in fixtures is smelly :)

**Pull request merged**